chain/pruned_block_dispatcher_test.go

```diff
@@ -0,0 +1,593 @@
+		RefreshPeersTicker: ticker.NewForce(time.Hour),
+		AllowSelfPeerConns: true,
+	})
+	require.NoError(t, err)
+
+	return h
+}
+
+// start starts the PrunedBlockDispatcher and asserts that connections are made
+// to all available peers.
+func (h *networkBlockTestHarness) start() {
+	h.t.Helper()
+
+	err := h.Start()
+	require.NoError(h.t, err)
+
+	h.peerMtx.Lock()
+	numPeers := len(h.peers)
+	h.peerMtx.Unlock()
+
+	for i := 0; i < numPeers; i++ {
+		h.assertPeerDialed()
+	}
+}
+
+// stop stops the PrunedBlockDispatcher and asserts that all internal fields of
+// the harness have been properly consumed.
+func (h *networkBlockTestHarness) stop() {
+	h.Stop()
+
+	select {
+	case <-h.dialedPeer:
+		h.t.Fatal("did not consume all dialedPeer signals")
+	default:
+	}
+
+	select {
+	case <-h.queriedPeer:
+		h.t.Fatal("did not consume all queriedPeer signals")
+	default:
+	}
+
+	require.Empty(h.t, h.blocksQueried)
+}
+
+// addPeer adds a new random peer available for use by the
+// PrunedBlockDispatcher.
+func (h *networkBlockTestHarness) addPeer() string {
+	addr := nextAddr()
+
+	h.peerMtx.Lock()
+	defer h.peerMtx.Unlock()
+
+	h.resetPeer(addr)
+	return addr
+}
+
+// resetPeer resets the internal peer connection state allowing the
+// PrunedBlockDispatcher to establish a mock connection to it.
+func (h *networkBlockTestHarness) resetPeer(addr string) {
+	h.peers[addr] = h.newPeer()
+
+	// Establish a mock connection between us and each peer.
+	inConn, outConn := pipe(
+		&conn{localAddr: addr, remoteAddr: "10.0.0.1:8333"},
+		&conn{localAddr: "10.0.0.1:8333", remoteAddr: addr},
+	)
+	h.localConns[addr] = outConn
+	h.remoteConns[addr] = inConn
+}
+
+// newPeer returns a new properly configured peer.Peer instance that will be
+// used by the PrunedBlockDispatcher.
+func (h *networkBlockTestHarness) newPeer() *peer.Peer {
+	return peer.NewInboundPeer(&peer.Config{
+		ChainParams:    &chaincfg.TestNet3Params,
+		DisableRelayTx: true,
+		Listeners: peer.MessageListeners{
+			OnGetData: func(p *peer.Peer, msg *wire.MsgGetData) {
+				go func() {
+					h.queriedPeer <- struct{}{}
+				}()
+
+				for _, inv := range msg.InvList {
+					// Invs should always be for blocks.
+					require.Equal(h.t, wire.InvTypeBlock, inv.Type)
+
+					// Invs should always be for known blocks.
+					block, ok := h.blocks[inv.Hash]
+					require.True(h.t, ok)
+
+					// Don't reply if requested.
+					if atomic.LoadUint32(&h.shouldReply) == 1 {
+						continue
+					}
+
+					go p.QueueMessage(block, nil)
+				}
+			},
+		},
+		Services:       wire.SFNodeNetwork | wire.SFNodeWitness,
+		AllowSelfConns: true,
+	})
+}
+
+// query requests the given blocks from the PrunedBlockDispatcher.
+func (h *networkBlockTestHarness) query(blocks []*chainhash.Hash,
+	opts ...query.QueryOption) (<-chan *wire.MsgBlock, <-chan error) {
+
+	h.t.Helper()
+
+	blockChan, errChan := h.Query(blocks, opts...)
+	select {
+	case err := <-errChan:
+		require.NoError(h.t, err)
```

Comment:
this select seems uneccessary as we don't expect it to happen?

Reply:
Within `PrunedBlockDispatcher.Query`, we have a call to `newRequest` which can return an error immediately. The select exists to handle that case.